File: src/main/java/maven/example/com/tools/archive/Archive.java
package maven.example.com.tools.archive;

import maven.example.com.tools.data.Data;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

public class Archive {
    private List<Data> datas;

    public Archive(List<Data> list) {
        this.datas = new ArrayList<>();
        datas.addAll(list);
    }

    public File filesToArchive(String nameArchive) throws IOException {
        File archive = new File(nameArchive);
        archive.createNewFile();
        FileOutputStream fos = new FileOutputStream(archive);
        ZipOutputStream zipOut = new ZipOutputStream(fos);

        // .txt файл в архив
        datas.forEach(data -> {
            try {
                File fileToZip = data.toFile();
                FileInputStream fis = new FileInputStream(data.toFile());
                ZipEntry zipEntry = new ZipEntry(fileToZip.getName());
                zipOut.putNextEntry(zipEntry);

                byte[] bytes = new byte[1024];
                int length;
                while ((length = fis.read(bytes)) >= 0) {
                    zipOut.write(bytes, 0, length);
                }

                fis.close();
                fileToZip.delete();
            } catch (Exception e) {
                System.err.println("Ошибка при помещения лога в архив!");
                e.printStackTrace();
            }
        });

        zipOut.closeEntry();
        zipOut.close();

        return archive;
    }
}

File: src/main/java/maven/example/com/utility/archive/Archive.kt
package maven.example.com.utility.archive

import maven.example.com.utility.data.Data
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.function.Consumer
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

class Archive {
    private var dates: MutableList<Data?>? = null

    constructor(list: MutableList<Data?>?) {
        this.dates = ArrayList()
        dates?.addAll(list?:throw NullPointerException())
    }

    @Throws(IOException::class)
    fun filesToArchive(nameArchive: String?): File {
        val archive = File(nameArchive?:throw NullPointerException())
        archive.createNewFile()
        val fos = FileOutputStream(archive)
        val zipOut = ZipOutputStream(fos)

        // .txt файл в архив
        dates?.forEach(Consumer { data: Data? ->
            try {
                val fileToZip = data!!.toFile()
                val fis = FileInputStream(data.toFile())
                val zipEntry = ZipEntry("${data.category}\\${fileToZip.name}")
                zipOut.putNextEntry(zipEntry)

                val bytes = ByteArray(1024)
                var length: Int
                while ((fis.read(bytes).also { length = it }) >= 0) {
                    zipOut.write(bytes, 0, length)
                }

                fis.close()
                fileToZip.delete()
            } catch (e: Exception) {
                System.err.println("Ошибка при помещения лога в архив!\n${e.stackTraceToString()}")
            }
        })

        zipOut.closeEntry()
        zipOut.close()

        return archive
    }
}